client/src/components/Filter.jsx

```diff
@@ -10,8 +24,208 @@ import React from "react";
  *
  * Cancel should clear any unapplied filters and close the dropdown
  */
-const Filter = () => {
-  return <></>;
+
+const Filter = ({ filterLabel }) => {
+  const [allFilterData, setAllFilterData] = useContext(FilterContext);
+  const [anchorEl, setAnchorEl] = useState(null);
+  const [icon, setIcon] = useState(<ArrowDropDown />);
+  const [buttonBgColor, setButtonBgColor] = useState("blue.200");
+  const [buttonBorderColor, setButtonBorderColor] = useState("gray.50");
+
+  const [checked, setChecked] = useState(allFilterData[filterLabel].data);
+  const [cancelButton, setCancelButton] = useState("");
+
+  /* buttonStatus
+    0 - all boxes are initially unchecked
+    1 - some or all boxes are initially checked
+    2 - some or all boxes are initially checked, and changes are made
+  */
+  const [buttonStatus, setButtonStatus] = useState(() => {
+    Object.keys(checked).forEach((key) => {
+      if (checked[key].checked) return 1;
+    });
+    return 0;
+  });
+  const handlePopUp = (isOpen, anchor = null) => {
+    if (isOpen) {
+      setAnchorEl(anchor);
+      setIcon(<Close />);
+      setButtonBgColor("red.300");
+      setButtonBorderColor("blue.200");
+    } else {
+      setAnchorEl(anchor);
+      setIcon(<ArrowDropDown />);
+      setButtonBgColor("blue.200");
+      setButtonBorderColor("gray.50");
+    }
+  };
+
+  const handleClick = (event) => {
+    let hasAtLeastOneChecked = false;
+    Object.keys(checked).forEach((key) => {
+      if (checked[key].checked) hasAtLeastOneChecked = true;
+    });
+
+    if (hasAtLeastOneChecked) changeButtonNames(1);
+    else changeButtonNames(0);
+    handlePopUp(true, event.currentTarget);
+  };
+
+  const handleClose = () => {
+    handlePopUp(false);
+    setChecked(allFilterData[filterLabel].data);
+  };
+
+  const handleChange = (event) => {
+    setChecked({
+      ...checked,
+      [event.target.name]: {
+        ...checked[event.target.name],
+        checked: event.target.checked,
+      },
+    });
+    if (buttonStatus > 0) changeButtonNames(2);
+  };
```

Comment:
![2023-04-18 13 55 02](https://user-images.githubusercontent.com/12107969/232904209-e6cb2322-969a-446e-a482-7a0f70c03037.gif)

If we turn on react dev tool's update painting, clicking on one checkbox looks like we're repainting a lot of the popover, meaning the whole component is updating. If we minimize the state updates, we might be able to clean this up as more of a performance change and keeping the logic a little more succinct.

Reply:
woww how do you do that

Reply:
![image](https://user-images.githubusercontent.com/12107969/233222759-38539554-e561-494e-91b3-2152a5a539f1.png)

https://react.dev/learn/react-developer-tools

browser extension!

Reply:
i've reduced the number of states but it still seems to look the same...